core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -410,6 +415,60 @@ impl HeaviestSubtreeForkChoice {
             .map(|(slot_hash, fork_info)| (slot_hash, fork_info.stake_voted_subtree))
     }
 
+    /// Dump the node `slot_hash_key` and propagate the stake subtraction up to the root of the
+    /// tree. Children of `slot_hash_key` are implicitely dumped as well, as if we were able to
+    /// chain them to a defunct parent that implies that they are defunct as well (consistent with
+    /// bank forks).
+    ///
+    /// Returns all removed slots
+    pub fn dump_node(&mut self, slot_hash_key: &SlotHashKey) -> Vec<Slot> {
+        let parent = {
+            let mut node_to_dump = self
+                .fork_infos
+                .get_mut(slot_hash_key)
+                .expect("Slot hash key must exist in tree");
+            // Remove stake to be aggregated up the tree
+            node_to_dump.stake_voted_subtree = 0;
+            node_to_dump.stake_voted_at = 0;
+            node_to_dump.is_dumped = true;
+            node_to_dump.parent
+        };
+
+        let mut update_operations: UpdateOperations = BTreeMap::new();
+        // Aggregate up to the root
+        self.insert_aggregate_operations(&mut update_operations, *slot_hash_key);
+        self.process_update_operations(update_operations);
+
+        // Remove node + all children
+        let mut to_visit = vec![*slot_hash_key];
+        let mut removed = Vec::new();
+
+        while !to_visit.is_empty() {
+            let current_node = to_visit.pop().unwrap();
+            println!("current node: {}", current_node.0);
+            let current_fork_info = self
+                .fork_infos
+                .remove(&current_node)
+                .expect("Node must exist in tree");
+
+            removed.push(current_node.0);
+            to_visit.extend(current_fork_info.children.into_iter());
+        }
+
+        if let Some(parent) = parent {
+            // Remove link from parent
+            let parent_fork_info = self
+                .fork_infos
+                .get_mut(&parent)
+                .expect("Parent must exist in fork infos");
+            parent_fork_info.children.remove(slot_hash_key);
+        } else {
+            warn!("Dumped root of tree {:?}", slot_hash_key);
+        }
+
```

Comment:
one thing to consider here is if it is necessary to update `latest_votes_per_pubkey`, *technically* a validator utilizing freebie would not have stake properly updated on the repaired version of the dropped block

Reply:
(freebie would be considered double vote here)

Reply:
actually this is already not an issue as long as the new vote hash is < old  nvm